plugins/net_plugin/net_plugin.cpp

```diff
@@ -4333,47 +4443,105 @@ namespace eosio {
+      auto [host, port, type] = split_host_port_type(peer_address);
+
+      auto resolver = std::make_shared<tcp::resolver>( my_impl->thread_pool.get_executor() );
+
+      resolver->async_resolve(host, port, 
+         [resolver, host = host, port = port, peer_address = peer_address, listen_address = listen_address, this]( const boost::system::error_code& err, const tcp::resolver::results_type& results ) {
+            connection_ptr c = std::make_shared<connection>( peer_address, listen_address );
+            c->set_heartbeat_timeout( heartbeat_timeout );
+            std::lock_guard g( connections_mtx );
+            auto [it, inserted] = connections.emplace( connection_details{
+               .host = peer_address,
+               .c = std::move(c),
+               .ips = results
+            });
+            if( !err ) {
+               it->c->connect( results, it );
+            } else {
+               fc_elog( logger, "Unable to resolve ${host}:${port} ${error}",
+                        ("host", host)("port", port)( "error", err.message() ) );
+               it->c->set_state(connection::connection_state::closed);
+               ++(it->c->consecutive_immediate_connection_close);
+            }
+      } );
+
       return "added connection";
    }
 
+   void connections_manager::update_connection_endpoint(connection_details_index::const_iterator it,
+                                                        const tcp::endpoint& endpoint) {
+      fc_dlog(logger, "updating connection endpoint");
+      std::unique_lock g( connections_mtx );
+      auto& index = connections.get<by_active_ip>();
+      index.modify_key(connections.project<by_active_ip>(it), [endpoint](tcp::endpoint& e) {
+         e = endpoint;
+      });
+   }
+
+   connections_manager::connection_details_index::iterator connections_manager::get_connection_iterator(const connection_ptr& c) {
+      std::lock_guard g( connections_mtx );
+      const auto& index = connections.get<by_connection>();
+      auto i = index.find(c);
+      return connections.project<by_host>(i);
+   }
+
    // called by API
    string connections_manager::disconnect( const string& host ) {
       std::lock_guard g( connections_mtx );
-      if( auto c = find_connection_i( host ) ) {
-         fc_ilog( logger, "disconnecting: ${cid}", ("cid", c->connection_id) );
-         c->close();
-         connections.erase(c);
+      auto& index = connections.get<by_host>();
+      if( auto i = index.find( host ); i != index.end() ) {
+         fc_ilog( logger, "disconnecting: ${cid}", ("cid", i->c->connection_id) );
+         i->c->close();
+         connections.erase(i);
          supplied_peers.erase(host);
          return "connection removed";
       }
       return "no known connection for host";
    }
 
    void connections_manager::close_all() {
-      fc_ilog( logger, "close all ${s} connections", ("s", connections.size()) );
+      auto& index = connections.get<by_host>();
```

Comment:
I think it would be better to grab the index after acquiring the lock.

Reply:
Revised.